File: src/main/java/sensormodels/activity/ActivitySensorData.java
package sensormodels.activity;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;
import database.LuceneManager;
import java.io.File;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.Date;
import org.apache.lucene.document.Document;
import org.apache.lucene.document.Field;
import org.apache.lucene.document.IntPoint;
import org.apache.lucene.document.StringField;
import org.bson.codecs.pojo.annotations.BsonIgnore;
import sensormodels.DatabaseModel;
import sensormodels.store.models.FileStoreModel;
import sensormodels.store.models.LuceneStoreModel;
import sensormodels.store.models.MongoStoreModel;
import sensormodels.store.models.MySqlStoreModel;
import utils.WebAppConstants;

public class ActivitySensorData implements MongoStoreModel, LuceneStoreModel, FileStoreModel,
    MySqlStoreModel {

  @BsonIgnore public static final String MY_SQL_TABLE_NAME = "ActivitySensorData";

  @BsonIgnore public static final String MONGO_COLLECTION_NAME = "ActivitySensorData";

  @BsonIgnore public static final String FILE_NAME = "Activity";

  @BsonIgnore private File file;

  @SerializedName("sensor_name")
  @Expose
  private String sensorName;

  @SerializedName("timestamp")
  @Expose
  private String timestamp;

  @SerializedName("time_stamp")
  @Expose
  private String timeStamp;

  @SerializedName("formatted_date")
  @Expose
  private String formattedDate;

  @SerializedName("sensor_data")
  @Expose
  private SensorData sensorData;

  @BsonIgnore
  public static String getMySqlTableName() {
    return MY_SQL_TABLE_NAME;
  }

  public String getSensorName() {
    return sensorName;
  }

  public void setSensorName(String sensorName) {
    this.sensorName = sensorName;
  }

  public String getTimestamp() {
    return timestamp;
  }

  public void setTimestamp(String timestamp) {
    this.timestamp = timestamp;
    setFormattedDate();
  }

  public String getTimeStamp() {
    return timeStamp;
  }

  public void setTimeStamp(String timeStamp) {
    this.timeStamp = timeStamp;
    this.formattedDate = WebAppConstants.inputDateFormat.format(new Date(timeStamp));
  }

  public String getFormattedDate() {
    return formattedDate;
  }

  public void setFormattedDate(String formattedDate) {
    this.formattedDate = formattedDate;
  }

  public void setFormattedDate() {
    if (timestamp != null) {
      this.formattedDate = WebAppConstants.inputDateFormat.format(new Date(timestamp));
    }
  }

  @Override
  public String getStartTime() {
    return this.getTimeStamp();
  }

  public SensorData getSensorData() {
    return sensorData;
  }

  public void setSensorData(SensorData sensorData) {
    this.sensorData = sensorData;
  }

  @Override
  @BsonIgnore
  public Document getDocument() {
    Document doc = new Document();
    doc.add(
        new IntPoint(
            LuceneManager.LuceneConstants.STEP_COUNT, this.getSensorData().getStepCounts()));
    doc.add(
        new IntPoint(
            LuceneManager.LuceneConstants.STEP_DELTA, this.getSensorData().getStepDelta()));
    doc.add(
        new StringField(
            LuceneManager.LuceneConstants.SENSOR_NAME, this.getSensorName(), Field.Store.YES));
    doc.add(
        new StringField(
            LuceneManager.LuceneConstants.FORMATTED_DATE, this.getSensorName(), Field.Store.YES));
    //         use a string field for timestamp because we don't want it tokenized
    doc.add(
        new StringField(
            LuceneManager.LuceneConstants.TIMESTAMP, this.getTimestamp(), Field.Store.YES));
    return doc;
  }

  @Override
  @BsonIgnore
  public String getMongoCollectionName() {
    return MONGO_COLLECTION_NAME;
  }

  @Override
  @BsonIgnore
  public Class<ActivitySensorData> getClassObject() {
    return ActivitySensorData.class;
  }

  @Override
  @BsonIgnore
  public String getTableName() {
    return MY_SQL_TABLE_NAME;
  }

  @Override
  @BsonIgnore
  public String getCreateTableQuery() {
    return "CREATE TABLE "
        + this.getTableName()
        + "(time_stamp VARCHAR(30) , "
        + " sensor_name CHAR(25) , "
        + " formatted_date CHAR(10) , "
        + " step_counts INTEGER, "
        + " step_delta INTEGER)";
  }

  @Override
  @BsonIgnore
  public String getInsertIntoTableQuery() {
    return " insert into "
        + this.getTableName()
        + " (time_stamp,formatted_date, sensor_name, step_counts,step_delta)"
        + " values (?, ?, ?, ?, ?)";
  }

  @Override
  @BsonIgnore
  public void fillQueryData(PreparedStatement preparedStmt) throws SQLException {
    preparedStmt.setString(1, this.getTimestamp());
    preparedStmt.setString(2, this.getFormattedDate());
    preparedStmt.setString(3, this.getSensorName());
    preparedStmt.setInt(4, this.getSensorData().getStepCounts());
    preparedStmt.setInt(5, this.getSensorData().getStepDelta());
  }

  @Override
  @BsonIgnore
  public String getFileName() {
    return FILE_NAME;
  }

  @Override
  @BsonIgnore
  public File getFile() {
    return this.file;
  }

  @Override
  @BsonIgnore
  public void setFile(File file) {
    this.file = file;
  }

  public static class SensorData {

    @SerializedName("step_counts")
    @Expose
    private Integer stepCounts;
    @SerializedName("step_delta")
    @Expose
    private Integer stepDelta;

    public SensorData() {}

    public Integer getStepCounts() {
      return stepCounts;
    }

    public void setStepCounts(Integer stepCounts) {
      this.stepCounts = stepCounts;
    }

    public Integer getStepDelta() {
      return stepDelta;
    }

    public void setStepDelta(Integer stepDelta) {
      this.stepDelta = stepDelta;
    }
  }
}

File: src/main/java/sensormodels/activity/ActivitySensorData.kt
package sensormodels.activity

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import database.LuceneManager
import org.apache.lucene.document.Document
import org.apache.lucene.document.Field
import org.apache.lucene.document.IntPoint
import org.apache.lucene.document.StringField
import org.bson.codecs.pojo.annotations.BsonIgnore
import sensormodels.store.models.SuperStoreModel
import utils.WebAppConstants.formatted
import java.io.File
import java.sql.PreparedStatement
import java.util.*

data class ActivitySensorData(
    override var file: File? = null,
    @SerializedName("sensor_name")
    @Expose
    var sensorName: String? = null,
    @SerializedName("timestamp")
    @Expose
    var time_stamp: String? = null,
    @SerializedName("formatted_date")
    @Expose
    var formattedDate: String? = null,
    @SerializedName("sensor_data")
    @Expose
    var sensorData: SensorData? = null,
) : SuperStoreModel {

    override val startTime: String?
        get() = timeStamp

    override val fileName: String = FILE_NAME

    @SerializedName("time_stamp")
    @Expose
    var timeStamp: String? = null
        set(timeStamp) {
            field = timeStamp
            formattedDate = Date(timeStamp).formatted()
        }

    fun setTimestamp(timestamp: String?) {
        this.time_stamp = timestamp
        setFormattedDate()
    }

    override fun setFormattedDate() {
        if (time_stamp != null) {
            formattedDate = Date(time_stamp).formatted()
        }
    }

    data class SensorData(
        @SerializedName("step_counts")
        @Expose
        var stepCounts: Int? = null,
        @SerializedName("step_delta")
        @Expose
        var stepDelta: Int? = null
    )

    companion object {
        @get:BsonIgnore
        @BsonIgnore
        val mySqlTableName = "ActivitySensorData"

        @BsonIgnore
        val MONGO_COLLECTION_NAME = "ActivitySensorData"

        @BsonIgnore
        val FILE_NAME = "Activity"
    }

    override val document: Document = Document().apply {
        add(
            IntPoint(
                LuceneManager.LuceneConstants.STEP_COUNT, sensorData!!.stepCounts!!
            )
        )
        add(
            IntPoint(
                LuceneManager.LuceneConstants.STEP_DELTA, sensorData!!.stepDelta!!
            )
        )
        add(
            StringField(
                LuceneManager.LuceneConstants.SENSOR_NAME, sensorName, Field.Store.YES
            )
        )
        add(
            StringField(
                LuceneManager.LuceneConstants.FORMATTED_DATE, sensorName, Field.Store.YES
            )
        )
        //         use a string field for timestamp because we don't want it tokenized
        add(
            StringField(
                LuceneManager.LuceneConstants.TIMESTAMP, time_stamp, Field.Store.YES
            )
        )
    }

    override fun fillQueryData(preparedStmt: PreparedStatement) {
        preparedStmt.apply {
            setString(1, time_stamp)
            setString(2, formattedDate)
            setString(3, sensorName)
            setInt(4, sensorData!!.stepCounts!!)
            setInt(5, sensorData!!.stepDelta!!)
        }
    }

    override val mongoCollectionName: String = MONGO_COLLECTION_NAME
    override val tableName: String = mySqlTableName
    override val createTableQuery: String =
        ("CREATE TABLE ${this.tableName}(time_stamp VARCHAR(30) ,  sensor_name CHAR(25) ,  formatted_date CHAR(10) ," +
                "  step_counts INTEGER,  step_delta INTEGER)")
    override val insertIntoTableQuery: String =
        (" insert into ${this.tableName} (time_stamp,formatted_date, sensor_name, step_counts,step_delta)" +
                " values (?, ?, ?, ?, ?)")
}